File: php/tests/PolymorphicArrayTest.php
<?php

declare(strict_types=1);

namespace Oojs\Tests;

use Oojs\ErrorCode;
use Oojs\Registry;
use PHPUnit\Framework\TestCase;

use function Oojs\validate;

/**
 * §7, §8 — Polymorphic dispatch via array.
 */
class PolymorphicArrayTest extends TestCase
{
    use TestHelpers;

    private Registry $r;
    private \Oojs\Model\Schema $schema;

    protected function setUp(): void
    {
        $this->r = new Registry();
        $this->r->loadDict([
            '$oojs' => '1.0',
            '$id'   => 'https://example.org/schemas/poly',
            'types' => [
                'Shape'  => ['abstract' => true, 'properties' => ['color' => ['type' => 'string']]],
                'Circle' => [
                    'extends'    => 'Shape',
                    'properties' => ['radius' => ['type' => 'number']],
                    'required'   => ['radius'],
                ],
                'Rect' => [
                    'extends'    => 'Shape',
                    'properties' => [
                        'width'  => ['type' => 'number'],
                        'height' => ['type' => 'number'],
                    ],
                    'required' => ['width', 'height'],
                ],
                'Canvas' => [
                    'properties' => ['shapes' => ['type' => 'array', 'items' => ['type' => 'Shape']]],
                ],
            ],
        ]);
        $this->schema = $this->r->getSchema('https://example.org/schemas/poly');
    }

    public function test_polymorphic_array_valid(): void
    {
        $instance = [
            '_type'  => 'Canvas',
            'shapes' => [
                ['_type' => 'Circle', 'radius' => 5.0],
                ['_type' => 'Rect', 'width' => 10.0, 'height' => 4.0],
            ],
        ];
        $errs = validate($instance, $this->schema->types['Canvas'], $this->schema, $this->r);
        $this->assertSame([], $errs);
    }

    public function test_polymorphic_array_abstract_item(): void
    {
        $instance = [
            '_type'  => 'Canvas',
            'shapes' => [['_type' => 'Shape', 'color' => 'red']],
        ];
        $errs = validate($instance, $this->schema->types['Canvas'], $this->schema, $this->r);
        $this->assertTrue($this->hasCode($errs, ErrorCode::ABSTRACT_TYPE));
    }

    public function test_polymorphic_array_missing_required(): void
    {
        $instance = [
            '_type'  => 'Canvas',
            'shapes' => [['_type' => 'Circle']], // missing radius
        ];
        $errs = validate($instance, $this->schema->types['Canvas'], $this->schema, $this->r);
        $this->assertTrue($this->hasCode($errs, ErrorCode::MISSING_REQUIRED));
    }
}

File: java/src/test/java/org/oojs/PolymorphicArrayTest.java
package org.oojs;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.oojs.model.Schema;

import java.util.List;

import static org.junit.jupiter.api.Assertions.*;

class PolymorphicArrayTest extends TestHelpers {

    private Registry r;
    private Schema schema;

    @BeforeEach void setUp() {
        r = new Registry();
        r.loadMap(map("$oojs", "1.0", "$id", "https://example.org/schemas/poly",
                "types", map(
                        "Shape", map("abstract", true, "properties", map("color", map("type", "string"))),
                        "Circle", map("extends", "Shape",
                                "properties", map("radius", map("type", "number")),
                                "required", list("radius")),
                        "Rect", map("extends", "Shape",
                                "properties", map("width", map("type", "number"), "height", map("type", "number")),
                                "required", list("width", "height")),
                        "Canvas", map("properties", map("shapes",
                                map("type", "array", "items", map("type", "Shape")))))));
        schema = r.getSchema("https://example.org/schemas/poly");
    }

    @Test void polymorphic_array_valid() {
        Object instance = map("_type", "Canvas", "shapes", list(
                map("_type", "Circle", "radius", 5.0),
                map("_type", "Rect", "width", 10.0, "height", 4.0)));
        assertEquals(List.of(), new Validator(r).validate(instance, schema.types.get("Canvas"), schema));
    }

    @Test void polymorphic_array_abstract_item() {
        Object instance = map("_type", "Canvas", "shapes", list(map("_type", "Shape", "color", "red")));
        assertTrue(hasCode(new Validator(r).validate(instance, schema.types.get("Canvas"), schema), ErrorCode.ABSTRACT_TYPE));
    }

    @Test void polymorphic_array_missing_required() {
        Object instance = map("_type", "Canvas", "shapes", list(map("_type", "Circle")));
        assertTrue(hasCode(new Validator(r).validate(instance, schema.types.get("Canvas"), schema), ErrorCode.MISSING_REQUIRED));
    }
}
